### PR TITLE
Some improvements from memblade branch

### PR DESCRIPTION
**Pull request by zhemao**
I'd like to merge in a few changes I made when working on the memblade paper that will be useful in general.

* Optimized NIC with split TX/RX IRQs and checksum offloading
* Set DRAM profiling interval in configuration
* Throttle outbound bandwidth from switch
* Allow packets to be split across batches in switch

### CODE REVIEW

File: target-design/switch/switch.cc [hunk 1]
```diff
@@ -11,7 +11,7 @@
 #include <omp.h>
 #include <cstdlib>
 
-#define IGNORE_PRINTF
+//#define IGNORE_PRINTF
```
Comment (sagark) [inline target-design/switch/switch.cc]: why is this commented out?

File: target-design/switch/switch.cc [hunk 2]
```diff
@@ -156,8 +163,8 @@ while (!pqueue.empty()) {
     switchpacket * tsp = pqueue.top().switchpack;
     pqueue.pop();
     uint16_t send_to_port = get_port_from_flit(tsp->dat[0], 0 /* junk remove arg */);
-    printf("packet for port: %x\n", send_to_port);
-    printf("packet timestamp: %ld\n", tsp->timestamp);
+    //printf("packet for port: %x\n", send_to_port);
+    //printf("packet timestamp: %ld\n", tsp->timestamp);
```
Comment (sagark) [inline target-design/switch/switch.cc]: why are these commented out?

**Comment by sagark** [comment]
how well tested is the new NIC code? have you re-run memcached/iperf/etc?

what's the perf improvement of the new NIC stuff?

**Comment by sagark** [comment]
Also while we're at it, can you take care of #74

File: target-design/switch/switch.cc (hunk 1 above)
Comment (zhemao) [reply target-design/switch/switch.cc]: Whoops. I thought I reverted those changes, but apparently not.

**Comment by zhemao** [comment]
I've run iperf3 with the new NIC code and it works pretty consistently.

I got up to 1.9 Gbps with the code at one point, but since updating rocket-chip I'm now getting just 1.6 Gbps.

**Comment by zhemao** [comment]
Fixed issue #74 and reverted the switch.cc printf changes.

**Comment by sagark** [comment]
@zhemao can you run the full memcached experiment on it and send me the graph?

**Comment by sagark** [comment]
@davidbiancolin does this conflict with your changes at all? if so we should merge your PR first since it's bigger.

**Comment by davidbiancolin** [comment]
The only conflict is firesim_top.cc is moved to firesim/firesim_top.cc i believe. There's nothing serious.

**Pull request closed without merging**